File: cordova-plugin-globeconnect/src/ios/Payment.swift
//
//  Payment.swift
//
//  Created by Rico Maglayon on 08/12/2016.
//

import Foundation

@objc(Payment) class Payment : CDVPlugin {
    var paymentAccessToken: String?
    var paymentAppId: String?
    var paymentAppSecret: String?
    var paymentAmount: Float?
    var paymentDescription: String?
    var paymentEndUserId: String?
    var paymentReferenceCode: String?
    var paymentTransactionOperationStatus: String?

    @objc(setAccessToken:) func setAccessToken(command: CDVInvokedUrlCommand) -> Payment {
        let argument = command.arguments[0] as? String ?? ""
        self.paymentAccessToken = argument

        return self
    }

    @objc(setAppId:) func setAppId(command: CDVInvokedUrlCommand) -> Payment {
        let argument = command.arguments[0] as? String ?? ""
        self.paymentAppId = argument

        return self
    }

    @objc(setAppSecret:) func setAppSecret(command: CDVInvokedUrlCommand) -> Payment {
        let argument = command.arguments[0] as? String ?? ""
        self.paymentAppSecret = argument

        return self
    }

    @objc(setAmount:) func setAmount(command: CDVInvokedUrlCommand) -> Payment {
        let argument = command.arguments[0] as? Float ?? 0.00
        self.paymentAmount = argument

        return self
    }

    @objc(setDescription:) func setDescription(command: CDVInvokedUrlCommand) -> Payment {
        let argument = command.arguments[0] as? String ?? ""
        self.paymentDescription = argument

        return self
    }

    @objc(setEndUserId:) func setEndUserId(command: CDVInvokedUrlCommand) -> Payment {
        let argument = command.arguments[0] as? String ?? ""
        self.paymentEndUserId = argument

        return self
    }

    @objc(setReferenceCode:) func setReferenceCode(command: CDVInvokedUrlCommand) -> Payment {
        let argument = command.arguments[0] as? String ?? ""
        self.paymentReferenceCode = argument

        return self
    }

    @objc(setTransactionOperationStatus:) func setTransactionOperationStatus(command: CDVInvokedUrlCommand) -> Payment {
        let argument = command.arguments[0] as? String ?? ""
        self.paymentTransactionOperationStatus = argument

        return self
    }

    @objc(sendPaymentRequest:) func sendPaymentRequest(command: CDVInvokedUrlCommand) -> Payment {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )

        // set the url
        let sendPaymentURL = "https://devapi.globelabs.com.ph/payment/v1/transactions/amount?access_token="+self.paymentAccessToken!

        // set the payload
        let data: [String: Any] = [
            "amount"                      : String(format: "%.2f", self.paymentAmount!),
            "description"                 : self.paymentDescription!,
            "endUserId"                   : self.paymentEndUserId!,
            "referenceCode"               : self.paymentReferenceCode!,
            "transactionOperationStatus"  : self.paymentTransactionOperationStatus!
        ]

        // set the header/s
        var headers = Dictionary<String, String>()
        headers["Content-Type"] = "application/json; charset=utf-8"

        // we need to convert first the payload to JSON
        do {
            // convert it!
            let jsonData = try JSONSerialization.data(
                withJSONObject: data,
                options: JSONSerialization.WritingOptions.prettyPrinted
            )

            // it is now in json so we need it to be a string so we can send it
            if let jsonPayload = String(data: jsonData, encoding: String.Encoding.utf8) {
                HTTPRequest(url: sendPaymentURL, headers: headers)
                    .post(
                        parameters: jsonPayload,
                        success: { data, _ in
                            DispatchQueue.global(qos: .utility).async {
                                DispatchQueue.main.async {
                                    pluginResult = CDVPluginResult(
                                        status: CDVCommandStatus_OK,
                                        messageAs: String(data: data, encoding: .utf8)
                                    )

                                    self.commandDelegate!.send(
                                        pluginResult,
                                        callbackId: command.callbackId
                                    )
                                }
                            }
                    }, failure: { error, data in
                        pluginResult = CDVPluginResult(
                            status: CDVCommandStatus_ERROR,
                            messageAs: String(data: data, encoding: .utf8)
                        )

                        self.commandDelegate!.send(
                            pluginResult,
                            callbackId: command.callbackId
                        )
                    })
            }
        } catch let error as NSError {
            // oops, error in converting it to JSON
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: error.localizedDescription
            )

            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        }

        return self
    }

    @objc(getLastReferenceCode:) func getLastReferenceCode(command: CDVInvokedUrlCommand) -> Payment {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )

        // set the url
        let getLastReferenceCodeURL = "https://devapi.globelabs.com.ph/payment/v1/transactions/getLastRefCode?app_id="+self.paymentAppId!+"&app_secret="+self.paymentAppSecret!

        // set the header/s
        var headers = Dictionary<String, String>()
        headers["Content-Type"] = "application/json; charset=utf-8"

        // perform request
        HTTPRequest(url: getLastReferenceCodeURL, headers: headers).get(
            success: { data, _ in
                DispatchQueue.global(qos: .utility).async {
                    DispatchQueue.main.async {
                        pluginResult = CDVPluginResult(
                            status: CDVCommandStatus_OK,
                            messageAs: String(data: data, encoding: .utf8)
                        )

                        self.commandDelegate!.send(
                            pluginResult,
                            callbackId: command.callbackId
                        )
                    }
                }
            },
            failure: { error, data in
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: String(data: data, encoding: .utf8)
                )

                self.commandDelegate!.send(
                    pluginResult,
                    callbackId: command.callbackId
                )
            })

        return self
    }
}

File: cordova-plugin-globeconnect/www/Payment.js
var exec = require('cordova/exec');

// Payment
var Payment = function(accessToken) {
    // short hand initialization
    if(!(this instanceof Payment)) {
        return new Payment(accessToken);
    }

    // class name
    this.class = 'Payment';

    // access token set?
    if(accessToken) {
        exec(null, null, this.class, 'setAccessToken', [accessToken]);
    }

    /**
     * Set app id.
     *
     * @param  string
     * @return this
     */
    this.setAppId = function(appId) {
        exec(null, null, this.class, 'setAppId', [appId]);

        return this;
    };

    /**
     * Set app secret.
     *
     * @param  string
     * @return this
     */
    this.setAppSecret = function(appSecret) {
        exec(null, null, this.class, 'setAppSecret', [appSecret]);

        return this;
    };

    /**
     * Set access token.
     *
     * @param  string
     * @return this
     */
    this.setAccessToken = function(accessToken) {
        exec(null, null, this.class, 'setAccessToken', [accessToken]);

        return this;
    };

    /**
     * Set amount.
     *
     * @param  double
     * @return this
     */
    this.setAmount = function(amount) {
        exec(null, null, this.class, 'setAmount', [amount]);

        return this;
    };

    /**
     * Set description.
     *
     * @param  string
     * @return this
     */
    this.setDescription = function(description) {
        exec(null, null, this.class, 'setDescription', [description]);

        return this;
    };

    /**
     * Set end user id.
     *
     * @param  string
     * @return this
     */
    this.setEndUserId = function(endUserId) {
        exec(null, null, this.class, 'setEndUserId', [endUserId]);

        return this;
    };

    /**
     * Set reference code.
     *
     * @param  string
     * @return this
     */
    this.setReferenceCode = function(appSecret) {
        exec(null, null, this.class, 'setReferenceCode', [appSecret]);

        return this;
    };

    /**
     * Set transaction operation status.
     *
     * @param  string
     * @return this
     */
    this.setTransactionOperationStatus = function(transactionOperationStatus) {
        exec(null, null, this.class, 'setTransactionOperationStatus', [transactionOperationStatus]);

        return this;
    };

    /**
     * Send payment request.
     *
     * @param  function
     * @param  function
     * @return this
     */
    this.sendPaymentRequest = function(successCallback, errorCallback) {
        var callback = function(data) {
            try {
                data = JSON.parse(data);
            } catch(e) {};

            successCallback.call(this, data);
        };

        exec(callback, errorCallback, this.class, 'sendPaymentRequest', []);

        return this;
    };

    /**
     * Get last reference code.
     *
     * @param  function
     * @param  function
     * @return this
     */
    this.getLastReferenceCode = function(successCallback, errorCallback) {
        var callback = function(data) {
            try {
                data = JSON.parse(data);
            } catch(e) {};

            successCallback.call(this, data);
        };

        exec(callback, errorCallback, this.class, 'getLastReferenceCode', []);

        return this;
    };
};

module.exports = Payment;
